File: day01/day01.go
package day01

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

func sumCalories(elfes []string) []int {
	calories := make([]int, len(elfes))
	for i, elfe := range elfes {
		for _, line := range strings.Split(elfe, "\n") {
			if line == "" {
				continue
			}
			cal, _ := strconv.Atoi(line)
			calories[i] += cal
		}
	}
	return calories
}

func Solve(input string, test bool, task int) (string, string) {
	res1 := ""
	res2 := ""

	elfes := strings.Split(input, "\n\n")
	calories := sumCalories(elfes)
	sort.Ints(calories)

	if task != 2 {
		maxCalories := calories[len(calories)-1]
		res1 = fmt.Sprintf("%d", maxCalories)
		fmt.Printf("Max calories: %d\n", maxCalories)
	}

	if task != 1 {
		topCalories := 0
		for _, cal := range calories[len(calories)-3:] {
			topCalories += cal
		}
		fmt.Printf("Top calories: %d\n", topCalories)
		res2 = fmt.Sprintf("%d", topCalories)
	}

	return res1, res2
}

File: aoc22_rust/src/day01.rs
use crate::types::Task;

fn sum_calories(elves: &Vec<&str>) -> Vec<i32> {
    elves
        .iter()
        .map(|elf| {
            elf.lines()
                .fold(0, |sum, line| sum + line.parse::<i32>().unwrap())
        })
        .collect()
}

pub fn solve(input: &str, _test: bool, task: Task) -> (String, String) {
    let mut res1 = "".to_string();
    let mut res2 = "".to_string();

    let elves = input.split("\n\n").collect();
    let mut calories = sum_calories(&elves);
    calories.sort();

    if !matches!(task, Task::Two) {
        let max_calories = calories.last().unwrap();
        res1 = max_calories.to_string();
    }
    if !matches!(task, Task::One) {
        let top_calories = calories.iter().rev().take(3).sum::<i32>();
        res2 = top_calories.to_string();
    }

    (res1, res2)
}
